### PR TITLE
Backport explicit casting

### PR DESCRIPTION
**Pull request by tenderlove**
This backports explicit casting to `change_column` in migrations.

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb [hunk 1]
```diff
@@ -1131,37 +1131,18 @@ def add_column(table_name, column_name, type, options = {})
...
-            rename_column(table_name, tmp_column_name, column_name)
-            commit_db_transaction
-          rescue
-            rollback_db_transaction
-          end
+        sql_type = type_to_sql(type, options[:limit], options[:precision], options[:scale])
+        sql = "ALTER TABLE #{quoted_table_name} ALTER COLUMN #{quote_column_name(column_name)} TYPE #{sql_type}"
+        sql << " USING #{options[:using]}" if options[:using]
+        if options[:cast_as]
+          sql << " USING CAST(#{quote_column_name(column_name)} AS #{type_to_sql(options[:cast_as], options[:limit], options[:precision], options[:scale])})"
```
Comment (jrafanie) [inline activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:1143]: @tenderlove is it valid to allow both USING and USING CAST?

**Comment by jrafanie** [comment]
@tenderlove Maybe you can squash the "oops" commit into "allow the USING" and "ugh" commit into "allow types" unless you're going for internet points :trollface:

File: activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb (hunk 1 above)
Comment (jrafanie) [inline activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:1143]: It's been a while, I really have no idea, it just looks funny.

Comment (tenderlove) [inline activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:1143]: No, the first case is if you want to provide custom SQL for casting.  The second case is if you want to use the CAST function (essentially what we were doing before).

**Comment by tenderlove** [comment]
@jrafanie Should I? The "oops" commits are in rails/master, and I just cherry picked them over to our branch. :cry:

File: activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb (hunk 1 above)
Comment (jrafanie) [inline activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:1143]: @tenderlove does it make sense to raise an exception if someone tries to use both using and cast_as or will PG raise due to invalid syntax?

**Comment by jrafanie** [comment]
@tenderlove No, I agree, if they come directly from rails master, we should cherry-pick them as is.  :laughing: 

I was under the impression you had to make some changes to make the commits work on 3.2. 

And the :trophy: for using the same "oops" commits multiple times goes to @tenderlove 

:clap:

**Pull request merged**